### ISSUE TITLE
HTTP/2 support

### ISSUE THREAD

**Issue by InterAl** [axios/axios#1175]
Is there any plan to add HTTP/2 support for node.js? Since it is now officially supported in node.js 9, I think it's highly desirable to have this library support it as well.

**Comment by IshamMohamed** [comment]
I would like to see this functionality.

**Comment by maxim25** [comment]
+1

**Comment by vgraveto** [comment]
+1

**Comment by nidrax69** [comment]
+1

**Comment by redfish31** [comment]
+1

**Comment by q-nishimura** [comment]
+1

**Comment by lyonish** [comment]
+1

**Comment by Khaledgarbaya** [comment]
Please use the emoji to vote, adding +1 comment won't help

**Comment by konstantinsteinmiller** [comment]
:+1:

**Comment by mashintsev** [comment]
👍

**Comment by AshishKirodian** [comment]
+1

**Comment by yumaojun03** [comment]
+1

**Comment by MykhailoMatiiasevych** [comment]
Since Node 10 is LTS and it has Stable HTTP/2 support I would like to have this functionality in axios

**Comment by qraynaud** [comment]
Looking at the actual code in axios, seems like a headache. Wouldn't it be easier to just pass in another transport that would be a wrapper for the http API over the htt2 API? You could write such a code more easily I think!

**Comment by qraynaud** [comment]
Should also be possible to write a simple adapter for this (look adapter in the configuration).

**Comment by range-of-motion** [comment]
So what's up with this? Are we fucked if we wanna use HTTP/2 with Axios? It's been a year, has anyone found a solution to this?

**Comment by snowdream** [comment]
Any Update?

**Comment by Y0geshsharma** [comment]
Hello guys,
 i faced an issue while using post for a file object.
"code snippet"
Axios
           .post(url,fileobject)
           .then()
           .catch(error=>{
console.log(error)
})
"

**Comment by gabrielmoncea** [comment]
Any update on this three year old feature?

**Comment by cb-manideep** [comment]
+1

**Comment by jasonsaayman** [comment]
Hi,

Thanks yes we would like to get to it in V1. I know that seems far off but for now we are trying to get back stability and then release new features.

Thanks

**Comment by rimiti** [comment]
Any update about this old request?

**Comment by snowdream** [comment]
Any update？

**Comment by dpikt** [comment]
You can subscribe to notifications on the issue for updates @snowdream.

**Comment by ahmedjafri09** [comment]
:(

**Comment by phoenix741** [comment]
Any new on this subject ?
I have an application in C++ that will listen only HTTP/2 and i am not able to communicate with it from nodejs.

**Comment by qraynaud** [comment]
> 
> 
> Any new on this subject ?
> I have an application in C++ that will listen only HTTP/2 and i am not able to communicate with it from nodejs.

@phoenix741 obviously you can using other libs or even the base node API. Though axios is very good, it’s not the alpha and the omega of HTTP clients in node!

**Comment by einSelbst** [comment]
> > Any new on this subject ?
> > I have an application in C++ that will listen only HTTP/2 and i am not able to communicate with it from nodejs.
> 
> @phoenix741 obviously you can using other libs or even the base node API. Though axios is very good, it’s not the alpha and the omega of HTTP clients in node!

@phoenix741 this chart might be helpful https://github.com/sindresorhus/got#comparison

**Comment by jasonsaayman** [comment]
Still planned, probably in an ESM re-write that will happen later this year. It is however on my list and radar.

**Comment by schoettler** [comment]
Any news?

**Comment by jasonsaayman** [comment]
Yeah, it is a work in progress. I will update when there is a pre-release

**Comment by rimiti** [comment]
Thanks @ lot @jasonsaayman. If you need some help, let us know!

**Comment by utsabc** [comment]
Hey, any updates on this? @jasonsaayman we are planning to use HTTP2 using Axios for our product

**Comment by mihir-mehta** [comment]
Hi Jay, 

Any update on the same ?

**Comment by victorvhpg** [comment]
@jasonsaayman any update?

**Comment by mohsenomidi** [comment]
Dear contributors 
Thanks for your great library 

I am exiting to hear about roadmap of http2 support, is there any estimation about this feature?

**Comment by mrnonz** [comment]
Still waiting, cheer up contributors!

**Comment by zzdhidden** [comment]
I custom the transport by `http2-wrapper` module for support HTTP/2.

```js

const http2 = require("http2-wrapper");
const axios = require("axios");
axios
  .get("https://httpbin.org/", {
    transport: {
      request: function request(options, handleResponse) {
        const req = http2.request(options, handleResponse);
        setImmediate(() => {
          // Remove the axios action `socket.setKeepAlive` because the HTTP/2 sockets should not be directly manipulated
          const listeners = req.listeners("socket");
          if (listeners.length) req.removeListener("socket", listeners[0]);
        });
        return req;
      },
    },
  })
  .then((res) => {
    console.log(res.data);
  });

```

**Comment by QBBAA** [comment]
Any news?

**Comment by pigping88** [comment]
node app client send a http request, the nginx server get the http1,
browser client send a http request, the nginx server get the http2,
now, the nginx only support http2, avoid the http1.
How to resolve ?

**Comment by mariuszste** [comment]
Some notes on the http2-wrapper solution by @zzdhidden

the `on('socket')` event will be called immediately if the socket is already present before we even had a chance to remove it. The only way to fix that issue is to not let it be registered in the first place.

can be done using something like this:
```js
const http2Transport = {
  request: function request(options, handleResponse) {
    const req = http2.request(options, handleResponse);

    const origOn = req.on.bind(req);
    req.on = (name, ...args) => {
      if (name != 'socket') {
        return origOn(name, ...args);
      }
    }
    return req;
  },
};
```

In the adapter version it's using `axios.defaults.adapter` which is no longer a function but now is an array of strings so that won't work. and i don't think there is an easy way to get the adapter out of axios really likes to hide their internals for some reason.

But yeah, official support for h2 would be awesome in axios.

**Comment by kaykyr** [comment]
Any update about this?

**Comment by 0xmaayan** [comment]
Any updates here?

**Comment by ryanhugh** [comment]
Any update here?

**Comment by maxim-kr-solidlogic** [comment]
Do you plan to update to http2?

**Comment by nestle49** [comment]
Any news?

**Comment by nestle49** [comment]
> Use fetch instead, I would strongly recommend not using axios in any new projects. See https://caniuse.com/fetch and for some examples https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

There is a problem of refactoring old projects. Why not just support http2?

**Comment by simllll** [comment]
> > Use fetch instead, I would strongly recommend not using axios in any new projects. See https://caniuse.com/fetch and for some examples https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
> 
> There is a problem of refactoring old projects. Why not just support http2?

Use a wrapper like https://www.npmjs.com/package/haxios which has the same interface as axios but uses fetch api behind the scene

**Comment by nestle49** [comment]
> > > Use fetch instead, I would strongly recommend not using axios in any new projects. See https://caniuse.com/fetch and for some examples https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
> > 
> > 
> > There is a problem of refactoring old projects. Why not just support http2?
> 
> Use a wrapper like https://www.npmjs.com/package/haxios which has the same interface as axios but uses fetch api behind the scene (disclaimer: haxios has been initially created by myself)

Okay, but support from axios is not planned? What is the reason for this?

**Comment by nestle49** [comment]
@jasonsaayman can u answer, please?

**Comment by nestle49** [comment]
Anybody else live here?

**Comment by mariuszste** [comment]
as someone said before: 
> Please use the emoji to vote, adding +1 comment won't help

Yes, i get this comment is also part of the problem, but most people seem to not read all the comments above before commenting.

The answer to the "is there any progress" is most likely no. the only "progress" I can find is #2521. And afaik http2 support has never been a priority for this project and probably never will be. So if you guys really want it you will have to make a PR yourself. Asking for progress won't make it faster.

**Comment by jasonsaayman** [comment]
i agree with @mariuszste, thing is many people try make large changes but end up stopping. might be due to lack of interest or maybe not realising the small HTTP library is actually quite complex at times. it may also be that internals are poorly documented and not even typed. however this is on the road map of ideas, dreams and aspirations. i may throw my hat in the ring on this some day too since i am ready to be more active again on the project. but if anyone wants to work on it please put in a pr and tag me, it would also be helpful to incrementally work on this as all in one go is normally just too much to review and very risky

**Comment by nestle49** [comment]
> i agree with @mariuszste, thing is many people try make large changes but end up stopping. might be due to lack of interest or maybe not realising the small HTTP library is actually quite complex at times. it may also be that internals are poorly documented and not even typed. however this is on the road map of ideas, dreams and aspirations. i may throw my hat in the ring on this some day too since i am ready to be more active again on the project. but if anyone wants to work on it please put in a pr and tag me, it would also be helpful to incrementally work on this as all in one go is normally just too much to review and very risky

It pleases and gives hope to our team. Unfortunately, the library, as you correctly noted, is too complex for us to make it ourselves

**Comment by jasonsaayman** [comment]
yeah my mission over the next while will be to try make the contribution much easier

**Comment by oleksiukmarydev** [comment]
> I custom the transport by `http2-wrapper` module for support HTTP/2.
> 
> ```js
> const http2 = require("http2-wrapper");
> const axios = require("axios");
> 
> const http2Transport = {
>   request: function request(options, handleResponse) {
>     const req = http2.request(options, handleResponse);
>     setImmediate(() => {
>       // Remove the axios action `socket.setKeepAlive` because the HTTP/2 sockets should not be directly manipulated
>       const listeners = req.listeners("socket");
>       if (listeners.length) req.removeListener("socket", listeners[0]);
>     });
>     return req;
>   },
> };
> 
> axios
>   .get("https://httpbin.org/", {
>     transport: http2Transport,
>   })
>   .then((res) => {
>     console.log(res.data);
>   });
> ```
> 
> adapter version
> 
> ```js
> function http2AdapterEnhancer(adapter) {
>   return async (config) => {
>     if (config.http2) {
>       let req;
>       config.transport = {
>         request: function request(options, handleResponse) {
>           req = http2.request(options, handleResponse);
>           return req;
>         },
>       };
>       const ret = adapter(config);
>       // Remove the axios action `socket.setKeepAlive` because the HTTP/2 sockets should not be directly manipulated
>       const listeners = req.listeners("socket");
>       if (listeners.length) req.removeListener("socket", listeners[0]);
>       return ret;
>     } else {
>       return adapter(config);
>     }
>   };
> }
> 
> const http2Adapter = http2AdapterEnhancer(axios.defaults.adapter);
> 
> axios
>   .get("https://httpbin.org/", {
>     http2: true,
>     adapter: http2Adapter,
>   })
>   .then((res) => {
>     console.log(res.data);
>   });
> ```

How do i use the transport? Is it still a working solution because i can't find any documentation on that. 
I've tried to create axios.js file as following but there are error that `https` and `./framer` cant be found.

```
const  http2 = require('http2-wrapper');
const axios = require('axios');
const http2Transport = {
  request: function request(options, handleResponse) {
    const req = http2.request(options, handleResponse);
    setImmediate(() => {
      // Remove the axios action `socket.setKeepAlive` because the HTTP/2 sockets should not be directly manipulated
      const listeners = req.listeners('socket');
      if (listeners.length) req.removeListener('socket', listeners[0]);
    });
    return req;
  },
};
const instance = axios.create();
instance.defaults.transport = http2Transport;
export default instance;
```

Should i add http2 transport separately from axios config somewhere?

**Comment by DaniilIsupov** [comment]
I'm waiting for HTTP/2. But I don't want to do PR

**Comment by nestle49** [comment]
Maybe I can donate to someone who can add http2 support?

**Comment by swcm-mnestler** [comment]
Looks like the work on HTTP/2 in https://github.com/nodejs/undici/pull/2061 is nearing an experimental release, a (native) `fetch` adapter (#1219, #5146) would get HTTP/2 support "for free". Very interested in this avenue

**Comment by simllll** [comment]
https://www.npmjs.com/package/haxios is something like this, it's using node-fetch behind the scene, has http2 support therefore, but imitates the axios interface, so it can be used as a drop in replacement. (disclaimer, haxios was invented by myself some time ago, but we are not using axios in any way anymore)

**Comment by nestle49** [comment]
any news?

**Comment by nestle49** [comment]
any updates?

**Comment by DaniilIsupov** [comment]
i would like to see HTTP/2 support already

**Comment by nestle49** [comment]
axios doesn't support http2 yet! :mad: :mad: 😠 😠 😡

**Comment by einSelbst** [comment]
guys, so many alternatives, haxios was mentioned, here is even more https://github.com/sindresorhus/got#comparison

**Comment by nestle49** [comment]
> guys, so many alternatives, haxios was mentioned

> This package has been deprecated

R u serious?

**Comment by nestle49** [comment]
galaxy far, far away...

**Comment by DaniilIsupov** [comment]
is there any news?

**Comment by alexandr-gerasimovich** [comment]
when will http2 support be available? it's no longer possible to live like this

**Comment by DaniilIsupov** [comment]
apparently `axios` is deprecated
i will migrate our entire ecosystem to `got` (there was support already in 2021)

**Comment by Tibing** [comment]
I've managed to build the http2 adapter available as an npm package, check it out here: https://github.com/uibakery/axios-http2-adapter

**Comment by nestle49** [comment]
@jasonsaayman Perhaps you could help at least decompose the problem or create a roadmap?

**Comment by nestle49** [comment]
@jasonsaayman

**Comment by DaniilIsupov** [comment]
I understand that this problem is with us forever

**Comment by jasonsaayman** [comment]
closing due to being very stale, if this is still relevant please open a new issue

**Comment by mariuszste** [comment]
Stale? There are like 500+ people waiting for this. What's the point of closing it when it's not implemented and there is clearly a LOT ([literally the most upvoted issue in the entire repo](https://github.com/axios/axios/issues?q=is%3Aissue+sort%3Areactions-%2B1-desc)) of interest? This is not stale by any stretch of the imagination, this should be on the top of the list judging by the activity and votes.

currently open issues asking for the same thing: #6066, #5192  
Should I "open a new issue" and make a third duplicate or what?

**Comment by jasonsaayman** [comment]
no need i will re-open just trying to see what needs to be attended to, but IMHO, to get this started I would encourage someone to start a PR to kick this off

**Comment by nestle49** [comment]
> no need i will re-open just trying to see what needs to be attended to, but IMHO, to get this started I would encourage someone to start a PR to kick this off

Maybe can u help to create roadmap? This is a very pressing problem, but many don’t even understand where to start PR.

**Comment by spyshower** [comment]
> I've managed to build the http2 adapter available as an npm package, check it out here: https://github.com/uibakery/axios-http2-adapter

@Tibing is this stable? Is it safe to use it in production?

**Comment by jasonsaayman** [comment]
@Tibing would you be comfortable working together to get this into our core?

**Comment by Tibing** [comment]
> @Tibing would you be comfortable working together to get this into our core?

I'm sorry, but I don't have free time for this right now

**Comment by Tibing** [comment]
> > I've managed to build the http2 adapter available as an npm package, check it out here: https://github.com/uibakery/axios-http2-adapter
> 
> @Tibing is this stable? Is it safe to use it in production?

I'm using it in production for a year already so everything should be fine

**Comment by nestle49** [comment]
+1

**Comment by wootwoot1234** [comment]
+1

**Comment by DaniilIsupov** [comment]
+1

**Comment by HakanSungur** [comment]
We are waiting for development. Http1 is dead for many. This is causing people using axios to abandon it. You should have prioritized and done this by now! We are still waiting!

**Comment by nestle49** [comment]
> We are waiting for development. Http1 is dead for many. 

axios dead for many too

**Comment by skilbjo** [comment]
yea? and what are you choosing instead? `node-fetch`?
nothing i've seen has as good an api as `axios`
there was `request.js` but that was deprecated in 2020, which is when i switched to `axios`

**Comment by simllll** [comment]
native fetch, [got](https://github.com/sindresorhus/got),[ ky](https://github.com/sindresorhus/ky), ..

but you can also still use axios, if you really want to use http2 on server side you could also try a wrapper like [gaxios](https://github.com/googleapis/gaxios).

**Comment by DaniilIsupov** [comment]
> yea? and what are you choosing instead? `node-fetch`? nothing i've seen has as good an api as `axios` there was `request.js` but that was deprecated in 2020, which is when i switched to `axios`

`ky` and `got` are great and convenient alternatives.

there is also an experimental version for `nodejs` without using `got`, which is described in the `undici` repository: https://github.com/nodejs/undici/issues/2750#issuecomment-1941009554

**Comment by nestle49** [comment]
Hello my dear friends

**Comment by jasonsaayman** [comment]
please see https://github.com/axios/axios/issues/6984 for a list of feature requests
